packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,102 @@
+import { getParentToChildPath, resolvePath } from "../parent"
+import { applyPatches, onPatches, OnPatchesDisposer, patchRecorder } from "../patch"
+import { isRootStore, registerRootStore, unregisterRootStore } from "../rootStore"
+import { clone } from "../snapshot"
+import { assertTweakedObject } from "../tweaker/core"
+
+/**
+ * Callback function for `DraftManager.withDraft`.
+ */
+export type WithDraftCallback<T extends object> = (draftNode: T) => boolean | undefined | void
+
+/**
+ * Manager class returned by `draftMiddleware` that allows you to make changes to a draft copy of
+ * the original subtree and apply them to the original subtree or reject them.
+ */
+export class DraftManager {
+  /**
+   * The draft copy of the original subtree.
+   */
+  private readonly subtreeRootDraft: object
+
+  /**
+   * The internal `onPatches` disposer.
+   */
+  private readonly disposer: OnPatchesDisposer
+
+  /**
+   * Creates an instance of `DraftManager`.
+   * Do not use directly, use `draftMiddleware` instead.
+   *
+   * @param subtreeRoot Subtree root target object.
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot)
+    if (isRootStore(subtreeRoot)) {
+      registerRootStore(this.subtreeRootDraft)
+    }
+    this.disposer = onPatches(subtreeRoot, patches => applyPatches(this.subtreeRootDraft, patches))
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is caled with a draft copy of `node`. Any changes made to the draft are
+   * applied to the original subtree when `fn` returns `true`. When `fn` returns `false`,
+   * `undefined` or nothing, the changes made to the draft are rejected.
+   */
+  public withDraft<T extends object>(node: T, fn: WithDraftCallback<T>): void {
```

Comment:
just wondering, if there any reason to take the selection of a node rather than just pass the whole object and do the selection inside the draft function if needed?

I don't think there would be any kind of speed penalty

Also that way it could make typing less error prone (DraftManager could have type T, which is passed from `draftManager` function)

Reply:
Basically what I mean is that it could be done as 
```
const dm = draftManager(something) // T is typeof something
dm.withDraft(draft => { // draft is WithDraftCallback of T above
  const selector = draft.a.b.c,
})
```

rather than
```
const dm = draftManager(something)
dm.withDraft(something.a.b.c, selector => {
})
```

and that way you automatically ensure you can't pass invalid children as first parameter to the function (since there´s no such first parameter).

Reply:
Ahh, but maybe you want it to make it easy to select a node on the draft given the original. If so nevermind :)

Reply:
> Ahh, but maybe you want it to make it easy to select a node on the draft given the original. If so nevermind :)

Yes, exactly. I'd provide the `DraftManager` instance to the children of the subtree root and select the draft of an original node from, e.g., a computed property or `onAttachedToRootStore`.